cpp/Cipher/MGLGenerateKeyPairInstaller.cpp

```diff
@@ -29,44 +32,78 @@ namespace margelo {
+          ecConfig = std::make_shared<EcKeyPairGenConfig>(
+            prepareEcKeyGenConfig(runtime, arguments));
+        } else {
+          throw std::runtime_error("KeyVariant not implemented"
+            + std::to_string((int)variant));
+        }
+
         auto promiseConstructor =
             runtime.global().getPropertyAsFunction(runtime, "Promise");
 
         auto promise = promiseConstructor.callAsConstructor(
             runtime,
             jsi::Function::createFromHostFunction(
-                runtime, jsi::PropNameID::forAscii(runtime, "executor"), 2,
-                [&jsCallInvoker, config](
+                runtime,
+                jsi::PropNameID::forAscii(runtime, "executor"),
+                4,
+                [&jsCallInvoker, variant, rsaConfig, ecConfig](
                     jsi::Runtime &runtime, const jsi::Value &,
                     const jsi::Value *promiseArgs, size_t) -> jsi::Value {
                   auto resolve =
                       std::make_shared<jsi::Value>(runtime, promiseArgs[0]);
                   auto reject =
                       std::make_shared<jsi::Value>(runtime, promiseArgs[1]);
 
-                  std::thread t([&runtime, resolve, reject,
-                                 jsCallInvoker, config]() {
+                  std::thread t([&runtime, resolve, reject, jsCallInvoker,
+                      variant, rsaConfig, ecConfig]() {
                     m.lock();
                     try {
-                      jsCallInvoker->invokeAsync([&runtime, config, resolve]() {
-                        auto keys = generateRSAKeyPair(runtime, config);
-                        auto publicKey = toJSI(runtime, keys.first);
-                        auto privateKey = toJSI(runtime, keys.second);
+                      jsCallInvoker->invokeAsync([&runtime, resolve,
+                          variant, rsaConfig, ecConfig]() {
+                        std::pair<jsi::Value, jsi::Value> keys;
+
+                        // switch on variant to get proper generateKeyPair
+                        if (variant == kvRSA_SSA_PKCS1_v1_5 ||
+                            variant == kvRSA_PSS ||
+                            variant == kvRSA_OAEP
+                        ) {
+                          keys = generateRsaKeyPair(runtime, rsaConfig);
+                        } else
+                        if (variant == kvEC) {
+                          keys = generateEcKeyPair(runtime, ecConfig);
+                        } else {
+                          throw std::runtime_error("KeyVariant not implemented"
+                            + std::to_string((int)variant));
+                        }
+
                         auto res = jsi::Array::createWithElements(
                           runtime,
                           jsi::Value::undefined(),
-                          publicKey,
-                          privateKey);
+                          keys.first,
+                          keys.second);
```

Comment:
In addition to the comment above about abstraction, we should probably also use the names `publicKey` and `privateKey` here again- this `.first` and `.second` thing is not a good idea imo as it's unreadable what this actually contains. Especially since both are `jsi::Value`s.